### ISSUE TITLE
HTTP Error 302 or 0Mbit upload

### ISSUE THREAD

**Issue by arnau1** [sivel/speedtest-cli#151]
Hello,
Since today I have a problem. Speedtest-cli fails with the error:
root@nagios:~# speedtest-cli
Retrieving speedtest.net configuration...
Retrieving speedtest.net server list...
Failed to retrieve list of speedtest.net servers:

 HTTP Error 302: The HTTP server returned a redirect error that would lead to an                                                                              infinite loop.
The last 30x error message was:
Found
HTTP Error 302: The HTTP server returned a redirect error that would lead to an                                                                              infinite loop.
The last 30x error message was:
Found

OR, sometimes, speedtest-cli reports me a normal result in "download speed" but with 0,00MBps of Upload Speed.

I've tried replacing http://c.speedtest.net for www, with no result.
The version of speedtest-cli is 0.3.3a and Python 2.7.1+

Can you help me please?

Thanks.

**Comment by sivel** [comment]
Does the no-https branch help with any of your issues?

https://github.com/sivel/speedtest-cli/blob/no-https/speedtest_cli.py

**Comment by rlex** [comment]
Faced same problem today. No-https version helped.

**Comment by sobkas** [comment]
Same problem on no-https:
./speedtest_cli.py
Retrieving speedtest.net configuration...
Retrieving speedtest.net server list...
Failed to retrieve list of speedtest.net servers:

 HTTP Error 302: The HTTP server returned a redirect error that would lead to an infinite loop.
The last 30x error message was:
Found
HTTP Error 302: The HTTP server returned a redirect error that would lead to an infinite loop.
The last 30x error message was:
Found

**Comment by mike-pt** [comment]
Been getting this recently but its somewhat random, I guess the 302 is expected just not sure why It ends up in a infinite loop.

Just got it now, but tried again and it worked...

Testing with curl doesn't indicate any 302
```
% curl -I http://www.speedtest.net/speedtest-servers-static.php                                                                                                                                                 
HTTP/1.1 200 OK
Cache-Control: maxage=600
Content-Type: text/xml
Date: Fri, 10 Jul 2015 16:57:30 GMT
Expires: Fri, 10 Jul 2015 17:07:30 GMT
Pragma: public
Server: Apache/2.2.25 (Unix) mod_ssl/2.2.25 OpenSSL/1.0.1e DAV/2 PHP/5.3.27
Vary: Accept-Encoding
X-Frame-Options: SAMEORIGIN
X-Powered-By: PHP/5.3.27
X-XSS-Protection: 1; mode=block
Content-Length: 0

% curl -I https://www.speedtest.net/speedtest-servers-static.php                                                                                                                                                
HTTP/1.1 200 OK
Cache-Control: maxage=600
Content-Type: text/xml
Date: Fri, 10 Jul 2015 16:57:36 GMT
Expires: Fri, 10 Jul 2015 17:07:36 GMT
Pragma: public
Server: Apache/2.2.25 (Unix) mod_ssl/2.2.25 OpenSSL/1.0.1e DAV/2 PHP/5.3.27
Vary: Accept-Encoding
X-Frame-Options: SAMEORIGIN
X-Powered-By: PHP/5.3.27
X-XSS-Protection: 1; mode=block
Content-Length: 0
```

And I get the same result for the config page, wonder if its maybe a cache issue server-side

**Comment by mrjonny2** [comment]
Similar issue for me on OS X Yosemite


Traceback (most recent call last):
  File "/usr/local/bin/speedtest", line 9, in <module>
    load_entry_point('speedtest-cli==0.3.1', 'console_scripts', 'speedtest')()
  File "/Library/Python/2.7/site-packages/speedtest_cli.py", line 686, in main
    speedtest()
  File "/Library/Python/2.7/site-packages/speedtest_cli.py", line 536, in speedtest
    servers = closestServers(config['client'])
  File "/Library/Python/2.7/site-packages/speedtest_cli.py", line 345, in closestServers
    uh = urlopen('http://www.speedtest.net/speedtest-servers-static.php')
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 127, in urlopen
    return _opener.open(url, data, timeout)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 410, in open
    response = meth(req, response)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 523, in http_response
    'http', request, response, code, msg, hdrs)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 442, in error
    result = self._call_chain(*args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 382, in _call_chain
    result = func(*args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 629, in http_error_302
    return self.parent.open(new, timeout=req.timeout)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 410, in open
    response = meth(req, response)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 523, in http_response
    'http', request, response, code, msg, hdrs)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 442, in error
    result = self._call_chain(*args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 382, in _call_chain
    result = func(*args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 629, in http_error_302
    return self.parent.open(new, timeout=req.timeout)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 410, in open
    response = meth(req, response)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 523, in http_response
    'http', request, response, code, msg, hdrs)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 442, in error
    result = self._call_chain(*args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 382, in _call_chain
    result = func(*args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 629, in http_error_302
    return self.parent.open(new, timeout=req.timeout)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 410, in open
    response = meth(req, response)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 523, in http_response
    'http', request, response, code, msg, hdrs)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 442, in error
    result = self._call_chain(*args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 382, in _call_chain
    result = func(*args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 629, in http_error_302
    return self.parent.open(new, timeout=req.timeout)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 410, in open
    response = meth(req, response)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 523, in http_response
    'http', request, response, code, msg, hdrs)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 442, in error
    result = self._call_chain(*args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 382, in _call_chain
    result = func(*args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 619, in http_error_302
    self.inf_msg + msg, headers, fp)
urllib2.HTTPError: HTTP Error 302: The HTTP server returned a redirect error that would lead to an infinite loop.
The last 30x error message was:
Found

**Comment by sivel** [comment]
The devel version has been pushed to master and released as v0.3.3.  Closing this issue.
